### PR TITLE
Row z-index and a column resizer double click callback

### PR DESCRIPTION
**Pull request by jeremyk**
## Description
Row z-index change: we remove the style= z-index so that it can be set with CSS based on a selected row class.
We added a callback for double clicking the resizer 
We added a showBottomBorder on all rows props.

## Motivation and Context
Row z-index changes:
- We wanted excel style drag down but the "handle" in the corner would be cut off by the row below. There was no easy way to fix that without making it so that the currently selected row would have a higher z-index. 
-In the process I removed what seemed to be an unnecessary wrapping component: FixedDataTableRowImpl but I can put it back if it had a purpose I didn't understand. My general impression is that FDT2 has more of these than necessary so I started moving in that direction.

Column resizer: so could allow the column to auto size to its contents (like excel).

ShowBottomBorder: we couldn't get the bottom border on the last row in short tables to show.

## How Has This Been Tested?
We are using this extensively in a private project that has a full QA team.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x ] My code follows the code style of this project.
- [x ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x ] I have read the **CONTRIBUTING** document.

NOTE: I have not updated the documentation and I did commit to the dist folder. I'm not really sure how we could use this ourselves easily if I didn't do that (maybe someone can explain it to me?). There are also a couple other changes that probably need to be reverted for this to get reintegrated which I am happy to do but thought I would submit this PR first to just find out if these changes are wanted before I go to that trouble. Thanks.

### CODE REVIEW

**Comment by wcjordan** [comment]
@KamranAsif could you take a look at this since you have the most extensive knowledge of issues related to row z-indexing?  Also if you have some time I'd appreciate a primer on what we've dealt with in the past there.

File: src/FixedDataTable.js [hunk 1]
```diff
@@ -335,6 +345,7 @@ var FixedDataTable = createReactClass({
       showScrollbarX: true,
       showScrollbarY: true,
       touchScrollEnabled: false,
+      showBorderOnAllRows: true,
```
Comment (wcjordan) [inline src/FixedDataTable.js]: This should be false by default

File: src/FixedDataTable.js [hunk 2]
```diff
@@ -709,8 +721,15 @@ var FixedDataTable = createReactClass({
     /*number|string*/ columnKey,
     /*object*/ event
   ) {
+
+    var nowMs = Date.now();
+    if (this.state.lastColumnResizeStart && nowMs - this.state.lastColumnResizeStart < 500) {
```
Comment (wcjordan) [inline src/FixedDataTable.js]: can you pull the < 500 into a double click timeout const in this file?

File: dist/fixed-data-table-base.css [hunk 3]
```diff
@@ -17,6 +17,107 @@
  * LICENSE file in the root directory of this source tree. An additional grant
  * of patent rights can be found in the PATENTS file in the same directory.
  *
```
Comment (jeremyk) [inline dist/fixed-data-table-base.css]: Yes

File: site/base.less [hunk 4]
```diff
@@ -6,7 +6,7 @@ html, body {
   -ms-text-size-adjust: 100%;
   margin: 0;
   padding: 0;
-  -webkit-font-smoothing: antialiased;
+  -webkit-font-smoothing: subpixel-antialiased;
```
Comment (jeremyk) [inline site/base.less]: I will revert

File: src/FixedDataTable.js (hunk 2 above)
Comment (jeremyk) [inline src/FixedDataTable.js]: Sure

File: src/FixedDataTable.js (hunk 1 above)
Comment (jeremyk) [inline src/FixedDataTable.js]: Yep. Will do

**Comment by wcjordan** [comment]
Kam's not available, so sorry for the wait.  I've added some comments and think we can accept something like this (although I may strip it out in favor of #60 & #61).  

I am seeing a lot of issues in the examples where headers & fixed columns are being overlapped by rows & scrollable columns when scrolling Y & X respectively.  I'm also seeing an issue with the Resizable Columns example no longer working.  Could you run npm run site-dev-server and fix the issues?

fixed_scroll_issue
header_issue

File: src/FixedDataTableRow.js [hunk 5]
```diff
@@ -240,62 +251,4 @@ class FixedDataTableRowImpl extends React.Component {
...
-  render() /*object*/ {
-    var style = {
-      width: this.props.width,
-      height: this.props.height,
-      zIndex: (this.props.zIndex ? this.props.zIndex : 0),
-    };
-    FixedDataTableTranslateDOMPosition(style, 0, this.props.offsetTop, this._initialRender);
-
-    return (
-      <div
```
Comment (jeremyk) [inline src/FixedDataTableRow.js]: Ok. I think I had a better reason but can't remember any more. I will put it back and see if my changes still work.

**Comment by wcjordan** [comment]
@jeremyk any update here?  I'm going to cut a release shortly and wondering if I should wait on this.

**Comment by jeremyk** [comment]
Sorry, been on vacation and busy. The main update is that the reason I removed the wrapping div is that the class needs to be added on the outer most div or it won't reorder the z-index of the row. I thought I noted that here but it seems that I did not.

Anyway, if you are still interested I can get this cleaned up in a day - just let me know. Thanks!

**Comment by wcjordan** [comment]
Sure just depends on your priorities (there will always be more releases!).  If you could give me an idea where you're at that will help me with deciding when to cut this weekend / Monday.

**Comment by jeremyk** [comment]
I will work on this tomorrow and if I don't finish I will give an update.

File: site/base.less (hunk 4 above)
Comment (jeremyk) [inline site/base.less]: BTW, I was looking at this some more and I kind of think this should just be removed all together: http://usabilitypost.com/2012/11/05/stop-fixing-font-smoothing/

Comment (wcjordan) [inline site/base.less]: I think it's worth filing an issue for it, but I don't think it relates to your changeset and alters the styling of the docs so I'd rather not block merge on time to research it.

If I'm missing something and it's related to the changeset though, let me know.

**Comment by jeremyk** [comment]
Ok, I made all the requested changes except restoring the wrapper as I need that for my z-index change. I also merged from upstream and resolved the conflicts.

**Comment by wcjordan** [comment]
Great thanks.  I'll give this a look soon and ship it out w/ a release this weekend or so.

Since I plan to do a minor version bump, the wrapper change should be less risky than doing it on a patch bump

File: src/FixedDataTable.js [hunk 6]
```diff
@@ -751,8 +764,15 @@ var FixedDataTable = createReactClass({
     /*number|string*/ columnKey,
     /*object*/ event
   ) {
+
+    var nowMs = Date.now();
+    if (this.state.lastColumnResizeStart && nowMs - this.state.lastColumnResizeStart < DOUBLE_CLICK_TIMEOUT) {
+      (this.props.onColumnResizeDoubleClick || emptyFunction)(columnKey);
+    }
```
Comment (KamranAsif) [inline src/FixedDataTable.js]: Theres no reason to re-invent the wheel here. You should have a separate callback passed that uses the native onDoubleClick event handler

File: src/FixedDataTableColumnResizeHandle.js [hunk 7]
```diff
@@ -141,9 +141,15 @@ var FixedDataTableColumnResizeHandle = createReactClass({
   },
 
   _onMove(/*number*/ deltaX) {
+
+    if (deltaX === 0) {
+      return;
```
Comment (KamranAsif) [inline src/FixedDataTableColumnResizeHandle.js]: Does this actually change anything?

File: src/FixedDataTableRow.js (hunk 5 above)
Comment (KamranAsif) [inline src/FixedDataTableRow.js]: +1 please revert

Comment (jeremyk) [inline src/FixedDataTableRow.js]: Below, I noted that "the reason I removed the wrapping div is that the class needs to be added on the outer most div or it won't reorder the z-index of the row."

File: src/FixedDataTable.js (hunk 6 above)
Comment (jeremyk) [inline src/FixedDataTable.js]: On a double click you were getting 2 regular resize callbacks. I wasn't trying to "reinvent the wheel" I just thought it easier to handle it all at this point.

File: src/FixedDataTableColumnResizeHandle.js (hunk 7 above)
Comment (jeremyk) [inline src/FixedDataTableColumnResizeHandle.js]: It basically stopped the resize callbacks when they were essentially noops. So it doesn't really do anything but I think it makes for a cleaner api.

**Comment by jeremyk** [comment]
I can certainly switch to an ondblclick and clean it up some more but I don't really want to keep working on this pull request if you guys don't really want it. My motivation is that we have a fork going and these are all the things that we need to switch back to the non-forked version. I think that is better for everyone but I don't really want to split it up for that reason.

**Comment by wcjordan** [comment]
We definitely do want to help merge you back to mainline, sorry if the feedback has come across as ungrateful.

The wrapping div issue may seem minor in your usage of FDT, but because FDT doesn't have a good system for styling components, a change like that can break the CSS styles of other users.  So we need to carefully investigate if there are alternatives or hold off on a change like that until a major release like the upcoming redux one.

Also have you had a chance to look into the screenshot I posted?

**Comment by jeremyk** [comment]
Ok, hope I didn't come across as if I don't want to help because I do. But I was thrown a little off guard as you asked for a bunch of changes and I did them all and then a bunch more were asked for. Anyway, I did see your screenshot and I fixed the issues with that. I could possibly move the classes I need to the wrapper but one of my main thoughts on FDT is that it has too many wrappers (there are 3 divs around each row) that don't seem to have a purpose (maybe I just don't understand the purpose) so I was trying to help removing one. Also it is set up to add all the classes to the inner wrapper so it seemed like a bunch of work to change how that wrapper was set up for minimal gain. I am adding a couple screenshots to help visualize this change.
after
before

**Comment by wcjordan** [comment]
Totally agree that there are unnecessary wrappers on divs, and we'd like to clean them up.  Unfortunately with libraries, even small things like that can break compatibility.

If you think you can make a pass at keeping it in place that will allow releasing this sooner.  Then filing a follow up ticket to reduce the wrappers can be targeted for the v1 release.

Either way let me know where you end up and I can adjust so it gets merged to the right place.
Thanks!

**Comment by wcjordan** [comment]
Closing this PR as stale.

**Pull request closed without merging**